spec.md

```diff
@@ -96,7 +96,60 @@ and eventually have it accepted as a Unicode Technical Standard (UTS).
 
 ## Data Model
 
-> _The data-only representation of a message._
+A Message provides the representation of a single message.
+It takes one of two forms,
+either as a PatternMessage or a SelectMessage.
+In a multi-locale context,
+a single message's choice of representation as one or the other
+should not affect its representation in other languages or locales.
+
+```ts
+type Message = PatternMessage | SelectMessage
+```
+
+A PatternMessage contains a list of PatternElement values,
+some of which are directly defined literal values,
+while others are placeholders with formatted values depending on additional data.
+
+```ts
+interface PatternMessage {
+  type: 'message'
+  value: PatternElement[]
+  meta?: Meta
+}
+```
+
+SelectMessage provides for the selection of one list of PatternElement values
+to use as the message's value when formatting,
+depending on the value of one or more Selector values.
+
+Each SelectCase is defined by a `key` of one or more string identifiers,
+and selection between them is made according to the corresponding Selector values.
+From this it follows that a valid SelectMessage must have at least as many `select` entries
+as its highest count of SelectCase `key` entries within its `cases`.
+
+As a SelectMessage is not a valid PatternElement,
+it is not possible for it to be placed within the `value` of another message.
+
+```ts
+interface SelectMessage {
+  type: 'select'
+  select: Selector[]
+  cases: SelectCase[]
+  meta?: Meta
+}
+
+interface Selector {
+  value: PatternElement
+  fallback?: string
```

Comment:
`suggestion`: can you add a sentence binding `fallback` string to `key` on the appropriate level making valid `SelectMessage` required to have all `fallback` strings be present in the `key` field.

It also is going to be a bit tricky to enforce:

```js
{
  type: 'select',
  select: [
    { value: ..., fallback: "masculine"}
    { value: ..., fallback: "few"}
  ],
  cases: [
    { value: ..., key: ["masculine", "one"] },
    { value: ..., key: ["masculine", "few"] },
    { value: ..., key: ["feminine", "one"] },
    { value: ..., key: ["feminine", "many"] },
  ],
  meta: null
}
```

In this scenario if we resolved first `select` to `feminine` then the default for the second is useless.

I'm wondering if we can do better, for example:



```js
{
  type: 'select',
  select: [
    { value: ..., fallback: "masculine"}
    { value: ..., fallback: "few"}
  ],
  defaults: {
    "masculine": "few",
    "feminine": "many",
  },
  cases: [
    { value: ..., key: ["masculine", "one"] },
    { value: ..., key: ["masculine", "few"] },
    { value: ..., key: ["feminine", "one"] },
    { value: ..., key: ["feminine", "many"] },
  ],
  meta: null
}
```

and validate that?

Or give up on partial fallbacks:


```js
{
  type: 'select',
  select: [
    { value: ..., fallback: "masculine"}
    { value: ..., fallback: "few"}
  ],
  cases: [
    { value: ..., key: ["masculine", "one"] },
    { value: ..., key: ["masculine", "few"], default: true },
    { value: ..., key: ["feminine", "one"] },
    { value: ..., key: ["feminine", "many"] },
  ],
  meta: null
}
```

Reply:
I don't it's ready to be submitted as a PR, but [here's my current draft](https://github.com/eemeli/message-format-wg/blob/select-messages/spec.md#select-case-resolution) of the select case resolution algorithm, which may provide relevant info on my views of the above. It probably makes more sense to first define at least the Formattables that it refers to.

In short, I think the `fallback` values _should_ rather than _must_ correspond to the case keys, though if they don't then it becomes more likely that the case selection fails, and we end up with an empty message. Which, to be clear, I think we should be able to handle, but strongly advise against.

And yes, that relationship ought to be made explicit in the spec.

The selection method I'd like to propose relies on case order providing an answer to the fallback vs. exact resolution. If the cases are ordered from more specific to less specific and keys are always checked from left to right, we can iterate through the cases and always select the first one that has keys which match either exactly or the fallback.

Reply:
`suggestion (n-b)`: Please consider adding at least a placeholder definition of `fallback`.

Reply:
Added a definition/description of `fallback`.

Reply:
Not completely addressed.
My mistake to pile up on the existing one, maybe a bad habit from Google Docs.
I can copy in a different thread if you prefer.

Reply:
@mihnita It is by far easiest for all concerned if a single thread here considers a single topic. I shall presume that the relevant part for this thread is the assertion that a selector such as your example may not be represented using the proposed data model. If there's another topic that you would also like to cover, please add a standalone comment for that.

If I understood your pseudocode right, it's a selector on the grammatical case of a variable called `genitive`, with a default value of `'nominative'`, yes? And that there are three cases, of which the third is the fallback. This is how it could be represented with the data model that's proposed here:

```js
{
  type: 'select',
  select: [
    {
      value: {
        type: 'function',
        func: 'grammatical_case',
        args: [
          { type: 'variable', var_path: [{ type: 'literal', value: 'genitive' }] }
        ],
        options: { default: { type: 'literal', value: 'nominative' } }
      },
      fallback: '???'
    }
  ],
  cases: [
    {
      key: ['nominative'],
      value: [
        { type: 'literal', value: "Hello user who's name is " },
        { type: 'variable', var_path: [{ type: 'literal', value: 'userName' }] },
        { type: 'literal', value: '!' }
      ]
    },
    {
      key: ['genitive'],
      value: [
        { type: 'literal', value: "Hello user who's name is " },
        {
          type: 'function',
          func: 'grammatical_case',
          args: [{ type: 'literal', value: 'userName' }],
          options: { grammatical_case: { type: 'literal', value: 'genitive' } }
        },
        { type: 'literal', value: '!' }
      ]
    },
    {
      key: ['???'],
      value: [{ type: 'literal', value: 'Hello!' }]
    }
  ]
}
```

Reply:
> single thread here considers a single topic

It was (is?) a single topic: "binding fallback string to key ..."
It is about how to represent that key: tag it at the top (in select), or on case (like Fluent does with `*`) or with a specially named key (like most programming languages)

----

> variable called genitive, with a default value of 'nominative', yes

No.

It is a fallback case that has nothing to do with any grammatical case.

It's saying _<< a correct Polish translation for "Hello {user}!" requires you to use vocative for the name. If you don't know how to form the vocative for that name at runtime, then fallback to "Hello!" >>_
So there is no user, no grammatical case, it is not a fallback to nominative.

Reply:
Then I don't think I've understood what your example selector means:
```
grammatical_case(genitive,  default: nominative)
```
That looks like pseudocode for some sort of function. Is that accurate, or is it meant to be something different? What are its inputs, and what does it produce? Is it somehow implicitly getting `user` as an additional parameter?

Reply:
It is pseudocode, I wanted to represent what you did in that json example.
But I didn't want to use it "as is" because I am not 100% sure what that mean.

Yes, it is a function, and yes, would need arguments.
I tried to leave out things are not relevant to the discussion.

Why would I invent some locale convention, in one message ("???"), instead a "universal" convention?

I understand that this is what Fluent does.
But a generic, clear "ultimate default" is what virtually all programming languages do.
So, they are wrong, and Fluent is right?
And if we want something programmer friendly, as stated many times before, shouldn't we be biased toward well established conventions, unless proven harmful?

In my opinion it has problems, and I explained what those are.

The only pro for the proposed approach is that if the fallback is one of the existing cases there is no need to repeat that message twice (for instance once for nominative, and once for `default`

Reply:
> It is pseudocode, I wanted to represent what you did in that json example.
> But I didn't want to use it "as is" because I am not 100% sure what that mean.

Are you perhaps referring to the data model examples that @zbraniecki posted earlier in this thread?

> Yes, it is a function, and yes, would need arguments.
> I tried to leave out things are not relevant to the discussion.

Perhaps these details are relevant, though? If we are to discuss whether a certain type of selection is possible or impossible, it would be good for us to establish exactly what it's doing. If you find it challenging to express in terms of function inputs and results, could you at least attempt to describe it in other words?

> Why would I invent some locale convention, in one message ("???"), instead a "universal" convention?

I wouldn't know, tbh. Clearly my initial read of your pseudocode was mistaken, that's why I'm asking for a clarficiation.

> I understand that this is what Fluent does.
> But a generic, clear "ultimate default" is what virtually all programming languages do.
> So, they are wrong, and Fluent is right?
> And if we want something programmer friendly, as stated many times before, shouldn't we be biased toward well established conventions, unless proven harmful?

Could you clarify how "Fluent is right" in this case? In its case selection, Fluent marks with a `*` one of the select cases to be the "ultimate default". This works well when selecting based on just one input, but it lacks precision when selecting on more than one input. Could it be that when talking of "Fluent" you mean the case selection method that's in the EZ proposal?

For MF2, we have previously reached a consensus decision to "Allow for selectors to select a case depending on the value of one or more input arguments." This is a different kind of selector than established conventions provide, and so we must by necessity find the best possible solution for our novel situation.

To see the need for assigning per-selector fallback values, consider the following scenario: A selector on two arguments, `$foo` and `$bar`. Then, let us presume that these cases are defined for them:
1. `$foo = fooX`, `$bar = barX`
2. `$foo = fooX`, `$bar = other`
3. `$foo = other`, `$bar = barX`
4. `$foo = other`, `$bar = other`

In other words, each of the cases matching `{fooX, other}` with `{barX, other}` are defined.

Now, what happens if we end up at runtime with (a) the values `$foo = fooX` and `$bar = ZZZ`? Or how about (b) the values `$foo = ZZZ` and `$bar = barX`?

If we define the fallback **per case** (as done by MF1, Fluent, and many others) we are effectively forced to select case 4 for both (a) and (b), because the arguments do not match any of the case keys.

If instead we define the fallback **per selector**, the case selection may select case 2 for (a) and case 3 for (b).

This is why the proposed data model includes the fallback per selector, because otherwise we cannot make the best choice in a situation like this.

Reply:
@mihnita Do you understand the specific scenario that I'm presenting in my previous comment, and how that requires a per-selector falback approach?

Reply:
I think I do.
And I don't think it requires per-selector default.

This can solve the same problem, with more flexibility:
```
switch (selector1(...), selector2(...), selector3(...)) {
    case [c1a, c2a, *] : ...
    case [c1b, *, c3a] : ...
    case [*, c2b, *] : ...
    case [*, *, *] : ...
}
```

Did you read the section about switch in the "whiteboard document"?
["6. Explicit default for selectors"](https://docs.google.com/document/d/1w6fxTfh4xeaqlKZ_QZK1sDpvTv1TqtBeh897aOT8qyk/edit?resourcekey=0-jRKq3bi_UzcnK-5gA8yjsw#heading=h.nkx529xpqnx)
Should I copy-paste it here?

Reply:
You proposal looks very interesting. Could you clarify exactly what the `*` signifies? Does it effectively correspond to the `other` fallback used by MF1, but on a per-selector basis?

> Did you read the section about switch in the "whiteboard document"? ["6. Explicit default for selectors"](https://docs.google.com/document/d/1w6fxTfh4xeaqlKZ_QZK1sDpvTv1TqtBeh897aOT8qyk/edit?resourcekey=0-jRKq3bi_UzcnK-5gA8yjsw#heading=h.nkx529xpqnx) Should I copy-paste it here?

Yes, I've read it, no need to copy-paste.

Reply:
`*` represents a `default`
It is not the same as `other`, which was "hijacked" by plural to mean something.

See description in the thread above:
```
So in the 1:1 with Stas one of the ideas was to have a special "default" case that is not "other", because that was already "hijacked" by plurals.
And that would be "outside any enums", like it is in most programming languages.
```

Reply:
Have I got it right that the logic would work something like this?

In the example you gave, we have three selectors, each with independent values. Let's call these values `s1`, `s2`, and `s3`. They're coming from your `selector#(...)` calls. Let's consider then the selection process and testing the first `case [c1a, c2a, *] : ...`:

If `s1 = c1a, s2 = foo, s3 = bar`, this would not be selected because `foo != c2a`.

With `s1 = c1a, s2 = c2a, s3 = bar`, this would be a valid selection, essentially because `* == x` for any `x`, including `bar`.

Have I understood your proposal right? I.e. that the data model should allow for a SelectCase key list to include a value representing this `*` default/fallback value?

Reply:
As requested by @mihnita on today's call, a note about the fallback values still being under discussion has been added to the PR.